ArduCopter/mode_zigzag.cpp

```diff
@@ -0,0 +1,373 @@
+
+    // update altitude target and call position controller
+    pos_control->set_alt_target_from_climb_rate_ff(target_climb_rate, G_Dt, false);
+    //adjusts target up or down using a climb rate
+
+    pos_control->update_z_controller();
+}
+
+//zigzag_has_arr_at_next_dest - judge if the plane is within a small area around the current destination
+bool Copter::ModeZigzag::zigzag_has_arr_at_dest()
+{
+    if (!zigzag_judge_moving.is_keeping_time) {
+        zigzag_judge_moving.is_keeping_time = true;
+        zigzag_judge_moving.last_judge_pos_time = AP_HAL::millis();
+        zigzag_judge_moving.last_pos = inertial_nav.get_position();
+    }
+    else {
+        if ((AP_HAL::millis() - zigzag_judge_moving.last_judge_pos_time) > 1000) {
+            Vector3f cur_pos = inertial_nav.get_position();
+            float dist_x = cur_pos.x - zigzag_judge_moving.last_pos.x;
+            float dist_y = cur_pos.y - zigzag_judge_moving.last_pos.y;
+            if ((dist_x*dist_x + dist_y*dist_y) < (ZIGZAG_RANGE_AS_STATIC * ZIGZAG_RANGE_AS_STATIC)) {
+                return true;
+            }
+            else {
+                zigzag_judge_moving.last_judge_pos_time = AP_HAL::millis();
+                zigzag_judge_moving.last_pos = inertial_nav.get_position();
+            }
+        }
+    }
+
+    return false;
+}
+
+// zigzag_calculate_next_dest - calculate next destination according to vector A-B and current position
+void Copter::ModeZigzag::zigzag_calculate_next_dest(Vector3f& next_dest, uint8_t next_A_or_B) const
+{
+    // calculate difference between A and B - vector AB and its direction
+    Vector3f pos_diff = zigzag_waypoint_state.B_pos - zigzag_waypoint_state.A_pos;
+    // get current position
+    Vector3f cur_pos = inertial_nav.get_position();
+    if (!zigzag_is_between_A_and_B) {
+        // if the drone's position is on the side of A or B
+        if (next_A_or_B == 0) {
+            next_dest = cur_pos + pos_diff;
+        }
+        if (next_A_or_B == 2) {
+            next_dest = cur_pos + pos_diff*(-1);
+        }
+    }
+    else {
+        // used to check if the drone is outside A-B scale
+        int next_dir = 1;
+        // if the drone's position is between A and B
+        float xa = zigzag_waypoint_state.A_pos.x;
+        float ya = zigzag_waypoint_state.A_pos.y;
+        float xb = zigzag_waypoint_state.B_pos.x;
+        float yb = zigzag_waypoint_state.B_pos.y;
+        float xc = cur_pos.x;
+        float yc = cur_pos.y;
+        next_dest.z = cur_pos.z;
+        if (next_A_or_B == 0) {
+            // calculate next B
+            Vector3f pos_diff_BC = cur_pos - zigzag_waypoint_state.B_pos;
+            if ((pos_diff_BC.x*pos_diff.x + pos_diff_BC.y*pos_diff.y) > 0) {
+                next_dir = -1;
+            }
+            float dis_from_AB = fabsf(xc*(yb - ya) / (xb - xa) - yc + (xb*ya - xa*yb) / (xb - xa)) / sqrtf(1 + ((yb - ya)*(yb - ya)) / ((xb - xa)*(xb - xa)));
```

Comment:
I have a white space between variables and operators
I think that there was better. I am curious that there is a white space.